File: grammar/grammar.go
package grammar

const (
	// The Magic Number
	Magic byte = 0x69

	// Pops the stack
	Pop byte = iota

	// Do nothing
	Noop

	// line++
	Advance

	// line--
	Back

	// Length of integer in unsigned leb128;
	// Followed by leb128
	Integer

	// String in bytes
	String

	// Null
	Null

	// name string
	GetVar

	// name string;
	// Expression on stack
	SetVar

	// name string;
	// Expression on stack
	GetInstance

	// numberOfArgs int;
	// [arg3, arg2, arg1, callee];
	// Will perform callee(arg1, arg2, arg3)
	Call

	// where int;
	// Expression on stack
	JumpIfFalse

	// where int
	Jump

	// numberOfParams int;
	// [param3, param2, param1, ...]
	// lengthOfCode int;
	// code
	Lambda

	// lengthOfCode int;
	// code
	Closure
)

File: src/grammar.rs
use strum::IntoEnumIterator;
use strum_macros::EnumIter;

use crate::grammar::Grammar::*;

#[derive(Debug, EnumIter, Copy, Clone)]
#[repr(u8)]
#[doc = "Bytecode grammar for Gorilla, ported from Go"]
pub enum Grammar {
    Magic = 0x69,
    Pop = 1,
    Noop,
    Advance,
    Back,
    Integer,
    String,
    Null,
    Getvar,
    Setvar,
    GetInstance,
    Call,
    JumpIfFalse,
    Jump,
    JumpTo,
    Function,
}

impl From<Grammar> for u8 {
    fn from(g: Grammar) -> Self {
        g as u8
    }
}

impl From<u8> for Grammar {
    fn from(x: u8) -> Self {
        for grm in Grammar::iter() {
            if grm as u8 == x {
                return grm;
            }
        }

        Noop
    }
}
